File: src/types/stats.ts
// System stats types - mirrors Rust models

export interface CpuStats {
  name: string;
  usage: number; // 0-100%
  frequency: number; // MHz
  cores: number;
  logical_cores: number;
  per_core_usage: number[];
  temperature?: number; // Celsius (from LibreHardwareMonitor sidecar)
  core_temperatures?: number[]; // Per-core temps (from sidecar)
  power?: number; // Watts (from sidecar)
}

export interface RamStats {
  total: number; // bytes
  used: number; // bytes
  available: number; // bytes
  usage_percent: number; // 0-100%
}

export interface GpuStats {
  name: string;
  usage: number; // 0-100%
  memory_total: number; // bytes
  memory_used: number; // bytes
  temperature?: number; // Celsius
  hot_spot_temperature?: number; // Celsius - GPU hottest point (from sidecar)
  fan_speed?: number; // 0-100%
  power?: number; // Watts (from sidecar)
  core_clock?: number; // MHz (from sidecar)
  memory_clock?: number; // MHz (from sidecar)
}

export interface SystemInfo {
  cpu_name: string;
  cpu_cores: number;
  cpu_threads: number;
  ram_total: number; // bytes
  gpu_name?: string;
  gpu_vram_total?: number; // bytes
  os_name: string;
  os_version: string;
  hostname: string;
  uptime_seconds: number;
}

export interface ProcessInfo {
  pid: number;
  name: string;
  cpu_usage: number; // 0-100%
  memory: number; // bytes
}

export interface SystemStats {
  cpu: CpuStats;
  ram: RamStats;
  gpu?: GpuStats;
  system_info: SystemInfo;
  processes: ProcessInfo[];
  timestamp: number;
}

// Sidecar status types - mirrors Rust SidecarStatusInfo

export type SidecarStatusType =
  | "not_started"
  | "running"
  | "stopped"
  | "error"
  | "requires_admin"
  | "binary_not_found";

export interface SidecarStatusPayload {
  status: SidecarStatusType;
  message?: string; // Error message when status is "error"
  restart_count: number;
  can_restart: boolean;
}

/**
 * Check if sidecar is in a healthy state
 */
export function isSidecarHealthy(status: SidecarStatusPayload): boolean {
  return status.status === "running";
}

/**
 * Check if sidecar has a recoverable error
 */
export function isSidecarRecoverable(status: SidecarStatusPayload): boolean {
  return status.status === "stopped" && status.can_restart;
}

/**
 * Get user-friendly message for sidecar status
 */
export function getSidecarStatusMessage(status: SidecarStatusPayload): string {
  switch (status.status) {
    case "not_started":
      return "Temperature monitoring initializing...";
    case "running":
      return "Temperature monitoring active";
    case "stopped":
      return status.can_restart
        ? `Temperature monitoring stopped. Restarting... (${status.restart_count}/3)`
        : "Temperature monitoring unavailable";
    case "requires_admin":
      return "Run as Administrator to enable temperature monitoring";
    case "binary_not_found":
      return "Temperature monitoring component not found";
    case "error":
      return status.message || "Temperature monitoring error";
    default:
      return "Unknown status";
  }
}

File: src-tauri/src/models/stats.rs
use serde::{Deserialize, Serialize};

/// Static system information (doesn't change frequently)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemInfo {
    pub cpu_name: String,
    pub cpu_cores: usize,
    pub cpu_threads: usize,
    pub ram_total: u64, // bytes
    pub gpu_name: Option<String>,
    pub gpu_vram_total: Option<u64>, // bytes
    pub os_name: String,
    pub os_version: String,
    pub hostname: String,
    pub uptime_seconds: u64,
}

/// Process information for top processes list
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessInfo {
    pub pid: u32,
    pub name: String,
    pub cpu_usage: f32, // 0-100%
    pub memory: u64,    // bytes
}

/// CPU statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CpuStats {
    pub name: String,
    pub usage: f32,     // 0-100%
    pub frequency: u64, // MHz
    pub cores: usize,
    pub logical_cores: usize,
    pub per_core_usage: Vec<f32>,
    pub temperature: Option<f32>, // Celsius (from WMI on Windows)
}

/// RAM/Memory statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RamStats {
    pub total: u64,         // bytes
    pub used: u64,          // bytes
    pub available: u64,     // bytes
    pub usage_percent: f32, // 0-100%
}

/// GPU statistics (optional - only available if GPU is detected)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GpuStats {
    pub name: String,
    pub usage: f32,               // 0-100%
    pub memory_total: u64,        // bytes
    pub memory_used: u64,         // bytes
    pub temperature: Option<u32>, // Celsius
    pub fan_speed: Option<u32>,   // 0-100%
}

/// Combined system statistics payload
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemStats {
    pub cpu: CpuStats,
    pub ram: RamStats,
    pub gpu: Option<GpuStats>,
    pub system_info: SystemInfo,
    pub processes: Vec<ProcessInfo>,
    pub timestamp: u64, // Unix timestamp in milliseconds
}

impl Default for CpuStats {
    fn default() -> Self {
        Self {
            name: String::from("Unknown CPU"),
            usage: 0.0,
            frequency: 0,
            cores: 0,
            logical_cores: 0,
            per_core_usage: Vec::new(),
            temperature: None,
        }
    }
}

impl Default for RamStats {
    fn default() -> Self {
        Self {
            total: 0,
            used: 0,
            available: 0,
            usage_percent: 0.0,
        }
    }
}

impl Default for GpuStats {
    fn default() -> Self {
        Self {
            name: String::from("Unknown GPU"),
            usage: 0.0,
            memory_total: 0,
            memory_used: 0,
            temperature: None,
            fan_speed: None,
        }
    }
}

impl Default for SystemInfo {
    fn default() -> Self {
        Self {
            cpu_name: String::from("Unknown CPU"),
            cpu_cores: 0,
            cpu_threads: 0,
            ram_total: 0,
            gpu_name: None,
            gpu_vram_total: None,
            os_name: String::from("Unknown"),
            os_version: String::new(),
            hostname: String::new(),
            uptime_seconds: 0,
        }
    }
}

impl Default for SystemStats {
    fn default() -> Self {
        Self {
            cpu: CpuStats::default(),
            ram: RamStats::default(),
            gpu: None,
            system_info: SystemInfo::default(),
            processes: Vec::new(),
            timestamp: 0,
        }
    }
}
